python/lsst/ts/observatory/control/auxtel/atcs.py

```diff
@@ -0,0 +1,1845 @@
+            ),
+        ]
+
+        for res in asyncio.as_completed(coro_list):
+            try:
+                await res
+            except Exception:
+                raise
+            else:
+                break
+            finally:
+                await self.cancel_not_done(coro_list)
+                self.check.atmcs = True
+                self.check.atdometrajectory = True
+
+    async def prepare_for_flatfield(self):
+        """A high level method to position the telescope and dome for flat
+        field operations.
+
+        The method will,
+
+            1 - disable ATDomeTrajectory
+            2 - send telescope to flat field position
+            3 - send dome to flat field position
+            4 - re-enable ATDomeTrajectory
+        """
+        await salobj.set_summary_state(self.rem.atdometrajectory, salobj.State.DISABLED)
+
+        await self.open_m1_cover()
+
+        atdometrajectory_check = self.check.atdometrajectory
+        self.check.atdometrajectory = False
+
+        try:
+            await self.point_azel(
+                target_name="FlatField position",
+                az=self.tel_flat_az,
+                el=self.tel_flat_el,
+                wait_dome=False,
+            )
+
+            try:
+                await self.stop_tracking()
+            except asyncio.TimeoutError:
+                self.log.debug("Timeout in stopping tracking. Continuing.")
+
+            await self.slew_dome_to(self.dome_flat_az)
+        except Exception as e:
+            self.check.atdometrajectory = atdometrajectory_check
+            raise e
+        finally:
+            await salobj.set_summary_state(
+                self.rem.atdometrajectory, salobj.State.ENABLED
+            )
+
+    async def stop_tracking(self):
+        """Task to stop telescope tracking."""
+
+        self.log.debug("Stop tracking.")
+
+        self.rem.atmcs.evt_atMountState.flush()
+        self.rem.atmcs.evt_allAxesInPosition.flush()
+
+        await self.rem.atptg.cmd_stopTracking.start(timeout=self.fast_timeout)
+
+        try:
+            at_mount_state = await self.rem.atmcs.evt_atMountState.aget(
+                timeout=self.long_timeout
+            )
+        except asyncio.TimeoutError:
+            self.log.warning("Timeout waiting for atMountState event.")
+        else:
+            # TODO: Remove hard coded 8 when real ATMCS is fixed.
+            while at_mount_state.state not in (8, ATMCS.AtMountState.TRACKINGDISABLED):
+                self.log.debug(
+                    f"Tracking state: {ATMCS.AtMountState(at_mount_state.state)!r}."
+                )
+                at_mount_state = await self.rem.atmcs.evt_atMountState.next(
+                    flush=False, timeout=self.long_timeout
+                )
+
+        try:
+            in_position = await self.rem.atmcs.evt_allAxesInPosition.aget(
+                timeout=self.fast_timeout
+            )
```

Comment:
I am flushing the event before sending the command...

Reply:
Oh. Good! I would gently suggest flushing here, rather than before sending the command because:
* It keeps the related code together -- a refactoring is less likely to lose the flush
* It is easier for the reader to find the flush.

(On the other hand: if you really only want to see values that result from the command then you'd have to flush first, hope that nobody else sent the same command, and then call next, not aget.)